---
title: "LAVA TUTORIAL"
author: "Josefin Werme, CTG Lab, VU Amsterdam"
date: "`r Sys.Date()`"
output: rmarkdown::github_document
vignette: >
  %\VignetteIndexEntry{tutorial}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

This tutorial shows you how to read in and analyse data with LAVA (**L**ocal **A**nalysis of [co]**V**ariant **A**ssociation): A tool developed for local genetic correlation (*r*~g~) analysis.

LAVA can analyse the local *r*~g~ between two or more phenotypes, analyse both binary and continuous phenotypes, and account for known or estimated sample overlap. It can also test the univariate local genetic signal for all phenotypes of interest (i.e. the local *h*^2^), which may be used to filter out non-associated loci, and model the local genetic relations between multiple phenotypes using conditional models.

The tutorial will show you how to install and run LAVA using some example input data. If you wish, you can inspect the data in the 'vignettes/data' folder.

------------------------------------------------------------------------

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Installing LAVA

To read in the genotype data, LAVA uses functions from the snpStats packge. As this package needs to be installed using BiocManager, you should do so before installing LAVA

```{r, eval=F}
install.packages("BiocManager")
BiocManager::install("snpStats")
```

You will also need devtools for the installation

```{r, eval=F}
install.packages("devtools")
```

LAVA can then be installed directly from github

```{r, eval=F}
devtools::install_github("https://github.com/josefin-werme/lava.git")
```

Or by downloading the code and installing it from a local directory

```{r, eval=F}
devtools::install("~/Programs/R/lava")		# adapt as necessary
```

```{r}
library(lava)
```

## Input format

*NOTE: Example input files can be found in the 'vignettes/data' directory*

#### As input, LAVA needs the following data:

-   **Reference genotype data** in plink format (.bim, .bed, .fam), used for the estimation of LD

    -   e.g. [1000 genomes](https://www.internationalgenome.org/data/), pre-processed input files can be found [here](%5Bhttps://ctg.cncr.nl/software/magma)](<https://ctg.cncr.nl/software/magma>))

-   **Input info file**, used for convenient processing of multiple phenotypes. Requires the columns:

    -   ***phenotype***: phenotype IDs

    -   ***cases***: number of cases

        -   set to 1 for continuous phenotypes

    -   ***controls:*** number of controls

        -   set to 0 for continuous phenotypes (important!); the case / control ratio is used to distinguish binary and continuous phenotypes, and is required for appropriate processing of binary phenotypes.

    -   ***prevalence*** (optional): the population prevalence of binary phenotypes

        -   this is only relevant if you want an estimate of the local population *h*^2^ for binary phenotypes. Estimates of the observed local sample *h*^2^ are still provided

    -   ***filename***: paths and file names to the relevant summary statistics

-   **Summary statistics** for all phenotypes of interest. To accommodate common summary statistics formats, a number of different column names are possible for the same column type (but **only ONE** header should be provided for each category; i.e do not provide a file that contains multiple valid SNP ID columns):

    -   ***SNP / ID / SNPID_UKB/ SNPID / MarkerName / RSID***: SNP IDs

    -   ***A1 / ALT***: effect allele

    -   ***A2 / REF***: reference allele

    -   ***N / NMISS / N_analyzed***: number of samples

    -   ***Z / T / STAT / Zscore***: if provided, no p-values or coefficients are needed; otherwise, please provide:

        -   ***B / BETA / OR / logOdds***: effect size coefficients

            -   please provide beta coefficients for continuous phenotypes, and odds ratios / log odds for binary phenotypes

        -   ***P***: p-values

-   **Locus definition file**: File that defines loci either based on genomic coordinates or a list of SNPs. Required headers:

    -   ***LOC***: locus ID

    -   ***CHR, START, STOP***: coordinates (i.e. basepair positions)

    -   ***SNPS***: list of SNPS. ';' separated, no white-space (optional)

        -   note: if a SNP list is provided, coordinates will be ignored and the reference data will be subsetted based on SNP IDs instead. This can be convenient if the locus definition file is based on a different GRChX version than the reference data. If no SNP list is provided, the versions of reference and the locus file MUST match!

-   **Sample overlap file** (optional)

    -   This can be obtained using the results from cross-trait LDSC (see the 'vignettes/sample_overlap.Rmd' file for a walk through on how to do this)

## Process input

Run the script below to process the sample input files. If you wish, you can examine the input object to get an idea of what your data looks like after processing. You can also check out the original data in the 'vignettes/data' folder.

```{r, results='hide'}
### Read in summarystatistics and related info
input = process.input(input.info.file="vignettes/data/input.info.txt",           # input info file
                      sample.overlap.file="vignettes/data/sample.overlap.txt",   # sample overlap file (can be set to NULL if there is no overlap)
                      ref.prefix="vignettes/data/g1000_test",                    # reference genotype data prefix
                      phenos=c("depression","neuro","bmi"))       # subset of phenotypes listed in the input info file that we want to process

# inspect the processed input data
ls(input)				      # this is actually an environment; hence ls() rather than str()
ls(input$sum.stats)   # processed summary statistics
input$info            # processed input info file, with additional variables N, prop_cases, binary computed by process.input()
input$sample.overlap  # sample overlap file
head(input$ref$bim)   # bim file from reference data
  
# read more about this function
?process.input()
```

**\>\> [ comment on the use of case control ratio to distinguish binary and continuous phentoypes ] \<\<**

## Prepare a locus for analysis

Before analysing the genetic correlation at a locus, we need to convert the marginal GWAS SNP effects within the locus to their corresponding joint effects (in order to account for the LD between SNPs), and estimate certain key parameter that are used for the analysis.

```{r}
### Read in locus info file
loci = read.loci("vignettes/data/test.loci")
head(loci)                          # inspect the locus file
?read.loci()                        # read more about the function and possible data formats

### Create a locus object for the first locus to prepare it for analysis
locus = process.locus(loci[1,], input)
```

If we inspect the locus object, we can see that it contains a lot of info about the locus, such as:

-   Locus coordinates ('chr','start','stop')

-   The number of SNPs ('n.snps') / PCs ('K') within the locus

-   The (estimated) PC projected joint SNP effects, $\delta$ ('delta')

-   The sampling covariance $\sigma^2$ ('sigma')

-   The genetic covariance matrix $\Omega = t(\delta)'\delta / K-\sigma^2$ ('omega') with corresponding correlation matrix $\Omega^*$ ('omega.cor')

```{r}

ls(locus)                               # inspect locus
c(locus$chr, locus$start, locus$stop)   # locus coordinates
str(locus$snps)                         # locus snps
locus$n.snps                            # N snps
locus$omega                             # genetic covariance matrix
locus$omega.cor                         # standardised genetic covariance matrix
locus$phenos							# locus phenotypes

?process.locus()  # find our more details in the function documentation
```

## Perform the univariate test

To determine the amount of local genetic signal for all phenotypes of interest, and filter out non-associated loci, we use the univariate test.

After obtaining the locus object, we pass it directly to the run.univ() function in order to test the local heritability for all phenotypes (or a subset)

```{r}
# for all phenotypes in the locus
run.univ(locus)

# or just a subset
run.univ(locus, phenos=c("depression","bmi"))
```

As there was little evidence of any genetic signal for either phenotypes, there will be no point in testing the local *r*~g~ (and if we did, the estimates would be unreliable and uninterpretable)..

If we check another locus however:

```{r}
locus = process.locus(loci[3,], input)
run.univ(locus)
```

we see that there is plenty of signal for all phenotypes, and we can proceed with the bivariate test!

## Analyse the bivariate local *r*~g~ between two phenotypes

To perform a bivariane *r*~g~ analysis, testing the local *r*~g~ between pairs of phenotypes, we simply pass the locus object the run.bivar() function

```{r}
run.bivar(locus)
```

When multiple phenotypes are entered simultaneously, the last phenotype will be treated as the target phenotype, and separate bivariate tests will be performed between this phenotype and all others. (Note that this test is symmetric, the order of the phenotypes doesn't matter!).

You can subset and change the order of the phenotypes using the 'phenos' argument:

```{r}
run.bivar(locus, phenos=c("neuro","bmi","depression"))
```

To filter automatically based on the univariate signal, you can also use the run.univ.bivar() function to perform both tests, with the condition that the bivariate test is only performed for the phenotypes that reach the univariate significance:

```{r}
# with the default p-value threshold of .05
run.univ.bivar(locus)

# or with a custom p-value threshold
run.univ.bivar(locus, univ.thresh = 1e-8)
```

If you would still like to print the bivariate output for unanalysed phentoypes, you can set the 'return.unanalysed' argument to TRUE.

```{r}
run.univ.bivar(locus, univ.thresh=1e-8, return.unanalysed=T)
```

## Apply the multivariate approaches to more than two phenotypes

For loci in which you find significant correlations between more than two phenotypes, you might want to apply one of two possible multivariate approaches in order to compute conditional local genetic correlations within the locus

### Multiple regression

With the multiple regression option, you can model the genetic signal for an outcome phenotype using several other predictor phenotypes simultaneously. When doing so, any potential local *r*~g~ between the predictor phenotypes will be accounted for, and their relation with the outcome will be conditional on all the other predictors.

For this analysis we will select a locus and a phenotype subset which has a bit more signal:

```{r}
input = process.input(input.info.file="vignettes/data/input.info.txt",
                      sample.overlap.file="vignettes/data/sample.overlap.txt",
                      ref.prefix="vignettes/data/g1000_test",
                      phenos=c("asthma","rheuma","diabetes","hypothyroidism"))

locus = process.locus(loci[19,], input)
```

Say that we are interested in modelling hypothyroidism as the outcome. Lets first check that there is actually some local genetic signal for all phenotypes, and that all predictors are genetically correlated with the outcome within this locus

```{r}
run.univ.bivar(locus)
```

They do! We can then run the multiple regression for all predictors.

Here, the last phenotype will again be treated as the outcome

```{r, eval=T}
run.multireg(locus)
```

Here you see that, by default, this function does not only return the full model with all predictors, but also all intermediate models. The reason for this is that when running a multiple regression it is important to be aware of how each predictor affects the fit of the other predictors in order to interpret the results correctly.

For example, in this case we can see that both asthma and diabetes are quite significant in the two-predictor models with rheuma, but not in the full model, or the two-predictor model with asthma and diabetes. This indicates that there is some collinearity between these predictors, which explains why neither appear very significant in the full model.

Had we not looked at the intermediate models beforehand, this pattern would have not been obvious. If we also didn't already know the bivariate correlations between the phenotypes, and only looked at the multivariate model, we might have made the erroneous conclusion that neither are relevant.

**\>\>\> [ comment on model r2 ] \<\<\<**

If for some reason you only want to return the full model, however, you can speed up the analysis by setting the 'only.full.model' argument to TRUE

```{r, eval=T}
run.multireg(locus, only.full.model=T)
```

### Partial correlation

In contrast with the multiple regression, the partial correlation allows you to compute the local r~*g*~ between two phenotypes of interest, conditioned on one or more other phenotypes.

For this example, we will use the same locus and set of phenotypes as before. Lets say we are now interested in the partial correlation between hypothyroidism and diabetes. Recall that there was also some collinearity between asthma and diabetes. We can therefore expect that if we compute the partial correlation between hypothyroidism and diabetes conditioned on asthma, that asthma might account for a notable proportion of the r~*g*~ between hypothyroidism and asthma (which we determined to be .86, from the bivariate test earlier)

```{r, eval=T}
run.partial.cor(locus, phenos=c("hypothyroidism","diabetes","asthma"))
```

Indeed, here you see that the partial correlation has been almost halved, and is no longer significant. The columns r2.phen1_z and r2.phen2_z confirm that asthma explains a substantial proportion of genetic variance for both hypothyroidism and diabetes.

If we instead condition on rheuma

```{r, eval=T}
run.partial.cor(locus, phenos=c("hypothyroidism","diabetes","rheuma"))
```

The partial correlation (.82) is now only slightly lower than the bivariate correlation (.86).

As seen previously, this phenotype was far less strongly genetically correlated with hypothyroidism (at .53), and as indicated by the r2's here, it is also explains less of the genetic variance in hypothyroidism and diabetes compared to asthma.

We can also condition on both asthma and rheuma (or any number of phenotypes) in one go:

```{r, eval=T}
run.partial.cor(locus, phenos=c("hypothyroidism","diabetes","rheuma","asthma"))
```

------------------------------------------------------------------------

## Example analysis script for running bivariate *r*~g~ analysis across all genomic loci

If you are interested in analysing the bivariate local *r*~g~'s between a large amount of phenotypes, we advice using a cluster computer. The example analysis script below shows how you might set up an R script that can be called from the command line, analysing the local *r*~g~'s across all loci in the locus file.

#### Bash

This is how you may call the R script from the command line

```{bash, eval=F}
Rscript "g1000_test" "test.loci" "input.info.file" "sample.overlap.file" "depression;bmi" "depression.bmi"
```

#### R script

```{r, eval=F}
# command line arguments, specifying input/output file names and phenotype subset
arg = commandArgs(T); ref.prefix = arg[1]; loc.file = arg[2]; info.file = arg[3]; sample.overlap.file = arg[4]; phenos = unlist(strsplit(arg[5],";")); out.fname = arg[6]

library(data.table); library(lava); print(paste("Running LAVA version",packageVersion("lava")))

### Read in data
loci = read.loci(loc.file); n.loc = nrow(loci)
input = process.input(info.file, sample.overlap.file, ref.prefix, phenos)

print(paste("Starting LAVA analysis for",n.loc,"loci"))
progress = ceiling(quantile(1:n.loc, seq(.05,1,.05)))   # (if you want to print the progress)

### Analyse
u=b=list()
for (i in 1:n.loc) {
        if (i %in% progress) print(paste("..",names(progress[which(progress==i)])))     # (printing progress)
        locus = process.locus(loci[i,], input)                                          # process locus
        
        # It is possible that the locus cannot be defined for various reasons (e.g. too few SNPs, negative variance), so the !is.null(locus) check is necessary before calling the analysis functions.
        # Note that it is also possible that individual phenotypes fail, in which case only the remaining phenotypes will be retured from the process.locus() function. 
        # If you require all phenotypes to be analysed (as we do here), also include the all(phenos %in% locus$phenos) condition in the if statement as well
        if (!is.null(locus) & all(phenos %in% locus$phenos)) {
                # extract some general locus info for the output
                loc.info = data.frame(locus = locus$id, chr = locus$chr, start = locus$start, stop = locus$stop, n.snps = locus$n.snps, n.pcs = locus$K)
                
                # run the univariate and bivariate tests
                loc.out = run.univ.bivar(locus, univ.thresh=1e-4)
                u[[i]] = cbind(loc.info, loc.out$univ)
                b[[i]] = cbind(loc.info, loc.out$bivar)
        }
}

# save the output
write.table(do.call(rbind,u), paste0(out.fname,".univ.lava"), row.names=F,quote=F,col.names=T)
write.table(do.call(rbind,b), paste0(out.fname,".bivar.lava"), row.names=F,quote=F,col.names=T)

print(paste0("Done! Analysis output written to ",out.fname,".*.lava"))
```
