<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Josefin Werme, CTG Lab, VU Amsterdam" />


<title>LAVA TUTORIAL</title>

<script src="data:application/javascript;base64,Ly8gUGFuZG9jIDIuOSBhZGRzIGF0dHJpYnV0ZXMgb24gYm90aCBoZWFkZXIgYW5kIGRpdi4gV2UgcmVtb3ZlIHRoZSBmb3JtZXIgKHRvCi8vIGJlIGNvbXBhdGlibGUgd2l0aCB0aGUgYmVoYXZpb3Igb2YgUGFuZG9jIDwgMi44KS4KZG9jdW1lbnQuYWRkRXZlbnRMaXN0ZW5lcignRE9NQ29udGVudExvYWRlZCcsIGZ1bmN0aW9uKGUpIHsKICB2YXIgaHMgPSBkb2N1bWVudC5xdWVyeVNlbGVjdG9yQWxsKCJkaXYuc2VjdGlvbltjbGFzcyo9J2xldmVsJ10gPiA6Zmlyc3QtY2hpbGQiKTsKICB2YXIgaSwgaCwgYTsKICBmb3IgKGkgPSAwOyBpIDwgaHMubGVuZ3RoOyBpKyspIHsKICAgIGggPSBoc1tpXTsKICAgIGlmICghL15oWzEtNl0kL2kudGVzdChoLnRhZ05hbWUpKSBjb250aW51ZTsgIC8vIGl0IHNob3VsZCBiZSBhIGhlYWRlciBoMS1oNgogICAgYSA9IGguYXR0cmlidXRlczsKICAgIHdoaWxlIChhLmxlbmd0aCA+IDApIGgucmVtb3ZlQXR0cmlidXRlKGFbMF0ubmFtZSk7CiAgfQp9KTsK"></script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<link rel="stylesheet" href="data:text/css,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" type="text/css" />




</head>

<body>




<h1 class="title toc-ignore">LAVA TUTORIAL</h1>
<h4 class="author">Josefin Werme, CTG Lab, VU Amsterdam</h4>
<h4 class="date">18th Dec, 2020</h4>



<p>This tutorial shows you how to read in and analyse data with LAVA (<strong>L</strong>ocal <strong>A</strong>nalysis of [co]<strong>V</strong>ariant <strong>A</strong>ssociation): A tool developed for local genetic correlation (<em>r<sub>g</sub></em>) analysis <strong>[paper ref]</strong>.</p>
<p>LAVA can analyse the local r<sub><em>g</em></sub> between two or more phenotypes, analyse both binary and continuous phenotypes, and account for known or estimated sample overlap. It can also test the univariate local genetic signal for all phenotypes of interest, which may be used to filter out unassociated loci.</p>
<p>The tutorial will show you how to install and run LAVA using some example input data. If you wish, you can inspect the data in the ‘vignettes/data’ folder.</p>
<hr />
<div id="installing-lava" class="section level2">
<h2>Installing LAVA</h2>
<p>To read in the genotype data, LAVA uses functions from the snpStats packge. As this package needs to be installed using BiocManager, its easiest to do so before installing LAVA</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;BiocManager&quot;</span>) <span class="co">#, repos=&#39;http://cran.rstudio.com/&#39;)</span></span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;BiocManager&quot;</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;snpStats&quot;</span>)</span></code></pre></div>
<p>You will also need devtools for the installation</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(<span class="st">&quot;devtools&quot;</span>)</span></code></pre></div>
<p>LAVA can then be installed directly from github</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;https://github.com/josefin-werme/lava.git&quot;</span>)   <span class="co"># adapt as necessary</span></span></code></pre></div>
<p>Or by downloading the code and installing it from a local directory</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;~/Programs/R/lava&quot;</span>)      <span class="co"># adapt as necessary</span></span></code></pre></div>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(lava)</span></code></pre></div>
</div>
<div id="input-format" class="section level2">
<h2>Input format</h2>
<p><em>NOTE: Example input files can be found in the ‘vignettes/data’ directory</em></p>
<div id="as-input-lava-needs-the-following-data" class="section level4">
<h4>As input, LAVA needs the following data:</h4>
<ul>
<li><p><strong>Reference genotype data</strong> in plink format (.bim, .bed, .fam), used for the estimation of LD</p>
<ul>
<li>e.g <a href="%5Bhttps://www.internationalgenome.org/data/">1000 genomes</a>](<a href="https://www.internationalgenome.org/data/" class="uri">https://www.internationalgenome.org/data/</a>)){.uri}</li>
</ul></li>
<li><p><strong>Input info file</strong>, used for convenient processing of multiple phenotypes. Requires the columns:</p>
<ul>
<li><p><em>‘phenotype’</em>: phenotype IDs</p></li>
<li><p><em>‘cases’</em>: number of cases</p>
<ul>
<li>set to 1 or the total sample size for continuous phenotypes</li>
</ul></li>
<li><p><em>‘controls’</em>: number of controls</p>
<ul>
<li>set to 0 for continuous phenotypes (important!)</li>
</ul></li>
<li><p><em>‘prevalence’</em> (optional)</p>
<ul>
<li>this is only relevant if you want an estimate of the local population h2 for binary phenotypes. Estimates of the sample h2 are still provided</li>
</ul></li>
<li><p><em>‘filename’</em>: paths and file names to the relevant summary statistics</p></li>
</ul></li>
<li><p><strong>Summary statistics</strong> for all phenotypes of interest. To accommodate common summary statistics formats, a number of different column names are possible for the same column type (but only ONE should be provided; i.e do not retain multiple SNP ID columns):</p>
<ul>
<li><p>‘SNP’ / ‘ID’ / ‘SNPID_UKB’/ ‘SNPID’ / ‘MarkerName’ / ‘RSID’: SNP IDs</p></li>
<li><p>‘A1’ / ‘ALT’: effect allele</p></li>
<li><p>‘A2’ / ‘REF’: reference allele</p></li>
<li><p>‘N’ / ‘NMISS’ / ‘N_analyzed’: number of samples</p></li>
<li><p>‘Z’ / ‘T’ / ‘STAT’ / ‘Zscore’: if provided, no p-values or coefficients are needed; otherwise, please provide:</p>
<ul>
<li><p>‘B’ / ‘BETA’ / ‘OR’ / ‘logOdds’: effect size coefficients</p>
<ul>
<li>please provide beta coefficients for continuous phenotypes, and odds ratios / log odds for binary phenotypes</li>
</ul></li>
<li><p>‘P’: p-values</p></li>
</ul></li>
</ul></li>
<li><p><strong>Locus definition file</strong>: File that defines loci either based on genomic coordinates or a list of SNPs. Required headers:</p>
<ul>
<li><p>‘LOC’: locus ID</p></li>
<li><p>‘CHR’, ‘START’, ‘STOP’: coordinates</p></li>
<li><p>‘SNPS’: list of SNPS (optional)</p>
<ul>
<li>note: if a SNP list is provided, coordinates will be ignored and the reference data will be subsetted based on SNP IDs instead. This can be convenient if the locus definition file is based on a different GRChX version than the reference data. If no SNP list is provided, the GRChX versions of reference and the locus file MUST match!</li>
</ul></li>
</ul></li>
<li><p><strong>Sample overlap file</strong> (optional)</p>
<ul>
<li>This can be obtained using the results from cross-trait LDSC (check out the ‘vignettes/sample_overlap.Rmd’ file for a walk through on how to do this)</li>
</ul></li>
</ul>
</div>
</div>
<div id="process-input" class="section level2">
<h2>Process input</h2>
<p>Run the script below to process the sample input files. If you wish, you can examine the input object to get an idea of what your data looks like after processing. You can also check out the original data in the ‘vingettes/data’ folder.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Navigate to the vignette directory</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">&quot;~/Programs/R/lava/vignettes&quot;</span>)    <span class="co"># adapt as necessary</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a><span class="do">### Read in summary statistics and related info</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>input <span class="ot">=</span> <span class="fu">process.input</span>(<span class="at">input.info.file=</span><span class="st">&quot;data/input.info.txt&quot;</span>,           <span class="co"># input info file</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>                      <span class="at">sample.overlap.file=</span><span class="st">&quot;data/sample.overlap.txt&quot;</span>,   <span class="co"># sample overlap file (can be set to NULL if there is no overlap)</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>                      <span class="at">ref.prefix=</span><span class="st">&quot;data/g1000_test&quot;</span>,                    <span class="co"># reference genotype data prefix</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">phenos=</span><span class="fu">c</span>(<span class="st">&quot;depression&quot;</span>,<span class="st">&quot;neuro&quot;</span>,<span class="st">&quot;bmi&quot;</span>))            <span class="co"># subset of phenotypes listed in the input info file that we want to process</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="co"># inspect the processed input data</span></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>(input)                     <span class="co"># this is actually an environment; hence ls() rather than str()</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>(input<span class="sc">$</span>sum.stats)   <span class="co"># processed summary statistics</span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>input<span class="sc">$</span>info            <span class="co"># processed input info file, with additional variables N, prop_cases, binary computed by process.input()</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>input<span class="sc">$</span>sample.overlap  <span class="co"># sample overlap file</span></span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(input<span class="sc">$</span>ref<span class="sc">$</span>bim)   <span class="co"># bim file from reference data</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a><span class="co"># read more about this function</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a>?<span class="fu">process.input</span>()</span></code></pre></div>
</div>
<div id="read-in-locus-definitions-and-process-a-locus" class="section level2">
<h2>Read in locus definitions and process a locus</h2>
<p>Before analysing the genetic correlation at a locus, we need to convert the marginal SNP effect sizes from GWAS to the corresponding joint effects in order to account for the LD in the locus. Only after we’ve done this, we can proceed with the analysis.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="do">### Read in locus info file</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>loci <span class="ot">=</span> <span class="fu">read.loci</span>(<span class="st">&quot;data/test.loci&quot;</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(loci)                          <span class="co"># inspect the locus file</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   LOC CHR     START      STOP</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 100   1 113418038 114664387</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 230   2  26894103  28819510</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3 266   2  57952946  59251996</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 4 374   2 191051955 193033982</span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 5 464   3  47588462  50387742</span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 6 950   6  25684630  26396200</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>?<span class="fu">read.loci</span>()                        <span class="co"># read more about the function and possible data formats</span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="do">### Create a locus object for the first locus to prepare it for analysis</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>locus <span class="ot">=</span> <span class="fu">process.locus</span>(loci[<span class="dv">1</span>,], input)</span></code></pre></div>
<p>If we inspect the locus object, we can see that it contains a lot of info about the locus, such as:</p>
<ul>
<li><p>Locus coordinates (‘chr’,‘start’,‘stop’)</p></li>
<li><p>The number of SNPs (‘N.snps’) / PCs (‘K’) within the locus</p></li>
<li><p>The (estimated) PC projected joint SNP effects, <span class="math inline">\(\delta\)</span> (‘delta’)</p></li>
<li><p>The sampling covariance <span class="math inline">\(\sigma^2\)</span> (‘sigma’)</p></li>
<li><p>The genetic covariance matrix <span class="math inline">\(\Omega = t(\delta)&#39;\delta / K-\sigma^2\)</span> (‘omega’) with corresponding correlation matrix <span class="math inline">\(\Omega^*\)</span> (‘omega.cor’)</p></li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ls</span>(locus)                               <span class="co"># inspect locus</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [1] &quot;binary&quot;    &quot;chr&quot;       &quot;delta&quot;     &quot;h2.latent&quot; &quot;h2.obs&quot;    &quot;id&quot;       </span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  [7] &quot;K&quot;         &quot;N&quot;         &quot;N.snps&quot;    &quot;omega&quot;     &quot;omega.cor&quot; &quot;phenos&quot;   </span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [13] &quot;sigma&quot;     &quot;snps&quot;      &quot;start&quot;     &quot;stop&quot;</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a><span class="fu">c</span>(locus<span class="sc">$</span>chr, locus<span class="sc">$</span>start, locus<span class="sc">$</span>stop)   <span class="co"># locus coordinates</span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1]         1 113418038 114664387</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(locus<span class="sc">$</span>snps)                         <span class="co"># locus snps</span></span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;  chr [1:2231] &quot;rs2360008&quot; &quot;rs1237670&quot; &quot;rs1235629&quot; &quot;rs1216539&quot; &quot;rs61819971&quot; ...</span></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>locus<span class="sc">$</span>N.snps                            <span class="co"># N snps</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1] 2231</span></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>locus<span class="sc">$</span>omega                             <span class="co"># genetic covariance matrix</span></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;               depression         neuro           bmi</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; depression  1.391995e-06 -2.074891e-07 -4.335870e-07</span></span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; neuro      -2.074891e-07  4.477136e-07 -2.674627e-07</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; bmi        -4.335870e-07 -2.674627e-07  7.443673e-07</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>locus<span class="sc">$</span>omega.cor                         <span class="co"># standardised genetic covariance matrix</span></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;            depression      neuro        bmi</span></span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; depression  1.0000000 -0.2628309 -0.4259550</span></span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; neuro      -0.2628309  1.0000000 -0.4633077</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; bmi        -0.4259550 -0.4633077  1.0000000</span></span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>locus<span class="sc">$</span>phenos                            <span class="co"># locus phenotypes</span></span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;      [,1]        </span></span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [1,] &quot;depression&quot;</span></span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [2,] &quot;neuro&quot;     </span></span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; [3,] &quot;bmi&quot;</span></span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>?<span class="fu">process.locus</span>()  <span class="co"># find our more details in the function documentation</span></span></code></pre></div>
</div>
<div id="perform-the-univariate-test" class="section level2">
<h2>Perform the univariate test</h2>
<p>If we determine the amount of local genetic signal for all phenotypes of interest, we can filter out unassociated loci and dramatically speed up the local r<sub><em>g</em></sub> analysis. This is done with the univariate test.</p>
<p>After obtaining the locus object, we pass it directly to the run.univ() function in order to test the local heritability for all phenotypes (or a subset)</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># for all phenotypes in the locus</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">run.univ</span>(locus)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         phen      h2.obs h2.latent          p</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 depression 8.05125e-05        NA 0.04240950</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      neuro 1.16406e-04        NA 0.03154340</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3        bmi 1.93535e-04        NA 0.00146622</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a><span class="co"># or just a subset</span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a><span class="fu">run.univ</span>(locus, <span class="at">phenos=</span><span class="fu">c</span>(<span class="st">&quot;depression&quot;</span>,<span class="st">&quot;bmi&quot;</span>))</span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         phen      h2.obs h2.latent          p</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 depression 8.05125e-05        NA 0.04240950</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2        bmi 1.93535e-04        NA 0.00146622</span></span></code></pre></div>
<p>As there was little evidence of any genetic signal for either phenotypes, and there will be little point in testing the local r<sub><em>g</em></sub>.</p>
<p>If we check another locus however:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>locus <span class="ot">=</span> <span class="fu">process.locus</span>(loci[<span class="dv">3</span>,], input)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">run.univ</span>(locus)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         phen      h2.obs h2.latent           p</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 depression 0.000260236        NA 4.13399e-07</span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      neuro 0.000573255        NA 1.25993e-14</span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3        bmi 0.000975242        NA 6.60429e-32</span></span></code></pre></div>
<p>there is plenty of signal for all phenotypes, and we can proceed with the bivariate test</p>
</div>
<div id="perform-a-bivariate-local-rg-analysis" class="section level2">
<h2>Perform a bivariate local r<sub><em>g</em></sub> analysis</h2>
<p>To proceed with the bivariane r<sub><em>g</em></sub> analysis, testing the local r<sub><em>g</em></sub> between pairs of phenotypes, we simply pass the locus object the run.bivar() function</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">run.bivar</span>(locus)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        phen1 phen2       rho rho.lower rho.upper         r2 r2.lower r2.upper</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 depression   bmi 0.0723039  -0.18619   0.32608 0.00522785        0  0.11006</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      neuro   bmi 0.1069490  -0.09842   0.31519 0.01143810        0  0.09935</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;          p</span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 0.568905</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 0.300006</span></span></code></pre></div>
<p>And if we look at the description of this function, we see that this too can take a phenotype vector, should only a subset of phenotypes be analysed. If multiple phenotypes are entered simultaneously, the last phenotype will be treated as the phenotype of interest, and separate bivariate tests will be performed between this phenotype and all others. (Note: this test is symmetric, which phenotype is considered the ‘outcome’ here doesn’t matter).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>?<span class="fu">run.bivar</span>()</span></code></pre></div>
<p>You can change the order of the phenotypes using the ‘phenos’ argument if you e.g. want to obtain the correlations between depression and all the other phenotypes</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">run.bivar</span>(locus, <span class="at">phenos=</span><span class="fu">c</span>(<span class="st">&quot;neuro&quot;</span>,<span class="st">&quot;bmi&quot;</span>,<span class="st">&quot;depression&quot;</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   phen1      phen2       rho rho.lower rho.upper         r2 r2.lower r2.upper</span></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 neuro depression 0.5278610   0.26848   0.77281 0.27863700  0.07208  0.59723</span></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2   bmi depression 0.0723039  -0.17911   0.32745 0.00522785  0.00000  0.10774</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;             p</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 0.000508409</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 0.566214000</span></span></code></pre></div>
<p>To filter automatically based on the univariate signal, you can also use the run.univ.bivar() function to perform both tests in one go, and only performs the bivariate test for the phenotypes that passed univariate significance</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># with the default p-value threshold of XXXX</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">run.univ.bivar</span>(locus)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $univ</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         phen      h2.obs h2.latent           p</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 depression 0.000260236        NA 4.13399e-07</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      neuro 0.000573255        NA 1.25993e-14</span></span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3        bmi 0.000975242        NA 6.60429e-32</span></span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $bivar</span></span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;        phen1 phen2       rho rho.lower rho.upper         r2 r2.lower r2.upper</span></span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 depression   bmi 0.0723039  -0.17882   0.32597 0.00522785        0  0.10886</span></span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      neuro   bmi 0.1069490  -0.10266   0.31140 0.01143810        0  0.09697</span></span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;          p</span></span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 0.567269</span></span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2 0.299425</span></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true" tabindex="-1"></a><span class="co"># or with a custom p-value threshold</span></span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true" tabindex="-1"></a><span class="fu">run.univ.bivar</span>(locus, <span class="at">univ.thresh =</span> <span class="fl">1e-8</span>)</span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $univ</span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;         phen      h2.obs h2.latent           p</span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 depression 0.000260236        NA 4.13399e-07</span></span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 2      neuro 0.000573255        NA 1.25993e-14</span></span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 3        bmi 0.000975242        NA 6.60429e-32</span></span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; </span></span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; $bivar</span></span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt;   phen1 phen2      rho rho.lower rho.upper        r2 r2.lower r2.upper        p</span></span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true" tabindex="-1"></a><span class="co">#&gt; 1 neuro   bmi 0.106949  -0.09913   0.30856 0.0114381        0  0.09523 0.301308</span></span></code></pre></div>
<p>as you can see, this function only proceeds with the bivariate analysis for all the phenotypes that reach univariate significance at the specified threshold.</p>
</div>
<div id="multivariate-approaches" class="section level2">
<h2>Multivariate approaches</h2>
<p>For loci in which you find significant correlations between more than two phenotypes, you might want to apply one of two possible multivariate approaches, which allow you to compute conditional local genetic correlations</p>
<div id="multiple-regression" class="section level3">
<h3>Multiple regression</h3>
<p>The multiple regression allows you to model the genetic signal for an outcome phenotype using several other predictor phenotypes simultaneously. When doing so, any potential local r~<em>g</em>~ between the predictor phenotypes will be accounted for, i.e., their genetic association with the outcome will be conditional on all the other predictors</p>
<p>** TODO : select another locus / phenotypes where there is more signal **</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="co"># For this analysis we will use a locus and phenotype subset in which there is a bit more signal</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="co">#locus = process.locus(loci[4,], input)</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">#run.multireg(locus)</span></span></code></pre></div>
</div>
<div id="partial-correlation" class="section level3">
<h3>Partial correlation</h3>
<hr />
</div>
</div>
<div id="example-analysis-script-for-running-bivariate-rg-analysis-across-all-genomic-loci" class="section level2">
<h2>Example analysis script for running bivariate r<sub><em>g</em></sub> analysis across all genomic loci</h2>
<p>If you are interested in e.g. the local bivariate genetic correlations between a large amount of phenotypes, we advice using a cluster computer. The example analysis script below shows how you might set up an R script that can be called from the command line</p>
<div id="bash-script" class="section level4">
<h4>Bash script</h4>
<div class="sourceCode" id="cb16"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="ex">Rscript</span> <span class="st">&quot;g1000_test&quot;</span> <span class="st">&quot;test.loci&quot;</span> <span class="st">&quot;input.info.file&quot;</span> <span class="st">&quot;sample.overlap.file&quot;</span> <span class="st">&quot;depression;bmi&quot;</span> <span class="st">&quot;depression.bmi&quot;</span></span></code></pre></div>
</div>
<div id="r-script" class="section level4">
<h4>R script</h4>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># command line arguments, specifying input/output file names and phenotype subset</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>arg <span class="ot">=</span> <span class="fu">commandArgs</span>(T); ref.prefix <span class="ot">=</span> arg[<span class="dv">1</span>]; loc.file <span class="ot">=</span> arg[<span class="dv">2</span>]; info.file <span class="ot">=</span> arg[<span class="dv">3</span>]; sample.overlap.file <span class="ot">=</span> arg[<span class="dv">4</span>]; phenos <span class="ot">=</span> <span class="fu">unlist</span>(<span class="fu">strsplit</span>(arg[<span class="dv">5</span>],<span class="st">&quot;;&quot;</span>)); out.fname <span class="ot">=</span> arg[<span class="dv">6</span>]</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table); <span class="fu">library</span>(lava); <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">&quot;Running LAVA version&quot;</span>,<span class="fu">packageVersion</span>(<span class="st">&quot;lava&quot;</span>)))</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="do">### Read in data</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>loci <span class="ot">=</span> <span class="fu">read.loci</span>(loc.file); n.loc <span class="ot">=</span> <span class="fu">nrow</span>(loci)</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>input <span class="ot">=</span> <span class="fu">process.input</span>(info.file, sample.overlap.file, ref.prefix, phenos)</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">&quot;Starting LAVA analysis for&quot;</span>,n.loc,<span class="st">&quot;loci&quot;</span>))</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>progress <span class="ot">=</span> <span class="fu">ceiling</span>(<span class="fu">quantile</span>(<span class="dv">1</span><span class="sc">:</span>n.loc, <span class="fu">seq</span>(.<span class="dv">05</span>,<span class="dv">1</span>,.<span class="dv">05</span>)))   <span class="co"># (if you want to print the progress)</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="do">### Analyse</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>u<span class="ot">=</span>b<span class="ot">=</span><span class="fu">list</span>()</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>n.loc) {</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (i <span class="sc">%in%</span> progress) <span class="fu">print</span>(<span class="fu">paste</span>(<span class="st">&quot;..&quot;</span>,<span class="fu">names</span>(progress[<span class="fu">which</span>(progress<span class="sc">==</span>i)])))     <span class="co"># (printing progress)</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>        locus <span class="ot">=</span> <span class="fu">process.locus</span>(loci[i,], input)                                          <span class="co"># process locus</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>        <span class="co"># It is possible that the locus cannot be defined for various reasons (e.g. too few SNPs, negative variance), so the !is.null(locus) check is necessary before calling the analysis functions.</span></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Note that it is also possible that individual phenotypes fail, in which case only the remaining phenotypes will be retured from the process.locus() function. </span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>        <span class="co"># If you require all phenotypes to be analysed, also include the all(phenos %in% locus$phenos) condition in the if statement as well</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (<span class="sc">!</span><span class="fu">is.null</span>(locus) <span class="sc">&amp;</span> <span class="fu">all</span>(phenos <span class="sc">%in%</span> locus<span class="sc">$</span>phenos)) {</span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>                <span class="co"># extract some general locus info for the output</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>                loc.info <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">locus =</span> locus<span class="sc">$</span>id, <span class="at">chr =</span> locus<span class="sc">$</span>chr, <span class="at">start =</span> locus<span class="sc">$</span>start, <span class="at">stop =</span> locus<span class="sc">$</span>stop, <span class="at">n.snps =</span> locus<span class="sc">$</span>N.snps, <span class="at">n.pcs =</span> locus<span class="sc">$</span>K)</span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>                </span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>                <span class="co"># run the univariate and bivariate tests</span></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>                loc.out <span class="ot">=</span> <span class="fu">run.univ.bivar</span>(locus, <span class="at">univ.thresh=</span><span class="fl">1e-4</span>)</span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a>                u[[i]] <span class="ot">=</span> <span class="fu">cbind</span>(loc.info, loc.out<span class="sc">$</span>univ)</span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>                b[[i]] <span class="ot">=</span> <span class="fu">cbind</span>(loc.info, loc.out<span class="sc">$</span>bivar)</span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a><span class="co"># save the output</span></span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">do.call</span>(rbind,u), <span class="fu">paste0</span>(out.fname,<span class="st">&quot;.univ.lava&quot;</span>), <span class="at">row.names=</span>F,<span class="at">quote=</span>F,<span class="at">col.names=</span>T)</span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a><span class="fu">write.table</span>(<span class="fu">do.call</span>(rbind,b), <span class="fu">paste0</span>(out.fname,<span class="st">&quot;.bivar.lava&quot;</span>), <span class="at">row.names=</span>F,<span class="at">quote=</span>F,<span class="at">col.names=</span>T)</span>
<span id="cb17-36"><a href="#cb17-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-37"><a href="#cb17-37" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">paste0</span>(<span class="st">&quot;Done! Analysis output written to &quot;</span>,out.fname,<span class="st">&quot;.*.lava&quot;</span>))</span></code></pre></div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
