<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Josefin Werme, CTG Lab, VU Amsterdam" />


<title>LAVA TUTORIAL</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#header {
text-align: center;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; }  code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">LAVA TUTORIAL</h1>
<h4 class="author">Josefin Werme, CTG Lab, VU Amsterdam</h4>
<h4 class="date">25th Sept, 2020</h4>



<p>This tutorial shows you how to read in and analyse data with LAVA (<strong>L</strong>ocal <strong>A</strong>nalysis of [co]<strong>V</strong>ariant <strong>A</strong>ssociation): A tool developed for the analysis of the local genetic correlation (<em>r<sub>g</sub></em>) between traits.</p>
<p>LAVA can analyse the local <em>r<sub>g</sub></em> between two or more phenotypes, analyse both binary and continuous phenotypes, and account for known or estimated sample overlap. In addition, it can test the univariate local genetic signal for all phenotypes of interest in order to filter out unassociated loci.</p>
<p>The tutorial contains example input data that we will use to demonstrate the relevant read-in and analysis functions. You can also inspect the data in the ‘lava/vignettes/data’ folder.</p>
<hr />
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co"># install.packages(&#39;devtools&#39;, repos=&#39;http://cran.rstudio.com/&#39;)</span></span>
<span id="cb1-2"><a href="#cb1-2"></a><span class="co"># library(devtools); devtools::install(&quot;../../lava&quot;)</span></span>
<span id="cb1-3"><a href="#cb1-3"></a><span class="kw">library</span>(lava)</span></code></pre></div>
<div id="input-format" class="section level2">
<h2>Input format</h2>
<div id="as-input-lava-needs-the-following-data" class="section level4">
<h4>As input, LAVA needs the following data:</h4>
<ul>
<li><p><strong>Reference genotype data</strong> in plink format (.bim, .bed, .fam), used for the estimation of LD</p></li>
<li><p><strong>Input info file</strong>, containing the columns:</p>
<ul>
<li><p><em>’</em>phenotype’: phenotype IDs</p></li>
<li><p>‘cases’: number of cases</p>
<ul>
<li>only relevant for binary phenotypes; set to 1, or the total sample size for continuous phenotypes</li>
<li>the number of cases / controls are used to calculate the case/control ratio for binary phenotypes, which is required in order to properly reconstruct the joint genetic effects for binary phenotypes</li>
</ul></li>
<li><p>‘controls’: number of controls</p>
<ul>
<li>set to 0 for continuous phenotypes</li>
</ul></li>
<li><p>‘filename’: path to summary statistics for all phenotypes</p></li>
</ul></li>
<li><p><strong>Summary statistics</strong> for all phenotypes of interest. To accomodate common summary statistics formats, a number of different column names are possible for the same column type (but only ONE should be provided; e.g do not enter multiple SNP ID columns):</p>
<ul>
<li><p><em>’</em>SNP’ / ‘ID’ / ‘SNPID_UKB’/ ‘SNPID’ / ‘MarkerName’ / ‘RSID’: SNP IDs</p></li>
<li><p><em>’</em>A1’: effect allele</p></li>
<li><p><em>’</em>A2’: reference allele</p></li>
<li><p><em>’</em>N’ / ‘NMISS’ / ‘N_analyzed’: number of samples</p></li>
<li><p><em>’</em>Z’ / ‘T’ / ‘STAT’ / ‘Zscore’: if provided, no p-values or coefficients are needed; otherwise, please provide:</p>
<ul>
<li><p>‘B’ / ‘BETA’ / ‘OR’ / ‘logOdds’: effect size coefficients.</p>
<ul>
<li>please provide beta coefficients for continuous phenotypes, and odds ratios / log odds for binary phenotypes</li>
</ul></li>
<li><p>‘P’: p-values</p></li>
</ul></li>
</ul></li>
<li><p><strong>Locus definition file</strong>: File that defines loci either based on genomic coordinates or a list of SNPs. Required headers:</p>
<ul>
<li><p>‘LOC’: locus ID</p></li>
<li><p>‘CHR’ + ‘START’ + ‘STOP’: coordinates</p></li>
<li><p>‘SNPS’: list of SNPS (optional)</p>
<ul>
<li><p>note: if a SNP list is provided, coordinates will be ignored, and loci will be subsetted based on SNPs instead. This can be convenient if the locus definition file is based on a different GRChX version than the reference data.</p>
<ul>
<li>if no SNP list is provided, the GRChX versions of reference and the locus file MUST match! (versions for the summary statistics don’t matter, since those are only subsetted based on SNP IDs)</li>
</ul></li>
</ul></li>
</ul></li>
<li><p><strong>Sample overlap file</strong> (if relevant)</p>
<ul>
<li>[<strong>TODO</strong>: add info on how to obtain this]</li>
</ul></li>
</ul>
</div>
</div>
<div id="process-input" class="section level2">
<h2>Process input</h2>
<p>Run the script below to process the sample input files. If you wish, you can examine the input object to get an idea of what your data looks like after processing. You can also check out the original data in the ‘lava/vingettes/data’ folder.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="co">### Input file paths</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>ref.fname =<span class="st"> &quot;data/g1000_test&quot;</span>                       <span class="co"># reference genotype data</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>input.info.file =<span class="st"> &quot;data/input.info.txt&quot;</span>             <span class="co"># input info file</span></span>
<span id="cb2-4"><a href="#cb2-4"></a>sample.overlap.file =<span class="st"> &quot;data/overlap.all.phenos.dat&quot;</span> <span class="co"># sample overlap file (can be set to NULL if there is no overlap)</span></span>
<span id="cb2-5"><a href="#cb2-5"></a></span>
<span id="cb2-6"><a href="#cb2-6"></a><span class="co">### Read in summary statistics and related info</span></span>
<span id="cb2-7"><a href="#cb2-7"></a><span class="co"># if only a subset of phenotypes are needed, set the phenos argument to a character vector indicating which ones</span></span>
<span id="cb2-8"><a href="#cb2-8"></a>input =<span class="st"> </span><span class="kw">process.input</span>(input.info.file, sample.overlap.file, ref.fname, </span>
<span id="cb2-9"><a href="#cb2-9"></a>                      <span class="dt">phenos=</span><span class="kw">c</span>(<span class="st">&quot;depression&quot;</span>,<span class="st">&quot;neuro&quot;</span>,<span class="st">&quot;bmi&quot;</span>))</span>
<span id="cb2-10"><a href="#cb2-10"></a></span>
<span id="cb2-11"><a href="#cb2-11"></a><span class="co"># inspect the processed input data</span></span>
<span id="cb2-12"><a href="#cb2-12"></a><span class="kw">str</span>(input<span class="op">$</span>sum.stats)  <span class="co"># processed summary statistics</span></span>
<span id="cb2-13"><a href="#cb2-13"></a>input<span class="op">$</span>info            <span class="co"># processed input info file, with additional variables N, prop_cases, binary computed by process.input()</span></span>
<span id="cb2-14"><a href="#cb2-14"></a>input<span class="op">$</span>sample.overlap  <span class="co"># sample overlap file</span></span>
<span id="cb2-15"><a href="#cb2-15"></a><span class="kw">head</span>(input<span class="op">$</span>ref<span class="op">$</span>bim)   <span class="co"># bim file from reference data</span></span>
<span id="cb2-16"><a href="#cb2-16"></a>  </span>
<span id="cb2-17"><a href="#cb2-17"></a><span class="co"># find out more about this function</span></span>
<span id="cb2-18"><a href="#cb2-18"></a>?<span class="kw">process.input</span>()</span></code></pre></div>
</div>
<div id="read-in-locus-info-and-create-a-locus-object" class="section level2">
<h2>Read in locus info and create a locus object</h2>
<p>Before analysing the genetic correlation at a locus, we need to convert the marginal SNP effect sizes from GWAS to the corresponding joint effects in order to account for the LD in the locus. Only after we’ve done this, we can proceed with the analysis.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="co">### Read in locus info file</span></span>
<span id="cb3-2"><a href="#cb3-2"></a>loci =<span class="st"> </span><span class="kw">read.loci</span>(<span class="st">&quot;data/test.loci&quot;</span>)  <span class="co"># read in locus file</span></span>
<span id="cb3-3"><a href="#cb3-3"></a><span class="kw">head</span>(loci)                          <span class="co"># inspect the locus file</span></span>
<span id="cb3-4"><a href="#cb3-4"></a><span class="co">#&gt;   LOC CHR     START      STOP</span></span>
<span id="cb3-5"><a href="#cb3-5"></a><span class="co">#&gt; 1 100   1 113418038 114664387</span></span>
<span id="cb3-6"><a href="#cb3-6"></a><span class="co">#&gt; 2 230   2  26894103  28819510</span></span>
<span id="cb3-7"><a href="#cb3-7"></a><span class="co">#&gt; 3 266   2  57952946  59251996</span></span>
<span id="cb3-8"><a href="#cb3-8"></a><span class="co">#&gt; 4 374   2 191051955 193033982</span></span>
<span id="cb3-9"><a href="#cb3-9"></a><span class="co">#&gt; 5 464   3  47588462  50387742</span></span>
<span id="cb3-10"><a href="#cb3-10"></a><span class="co">#&gt; 6 950   6  25684630  26396200</span></span>
<span id="cb3-11"><a href="#cb3-11"></a>?<span class="kw">read.loci</span>()                        <span class="co"># read more about the function and possible data formats</span></span>
<span id="cb3-12"><a href="#cb3-12"></a></span>
<span id="cb3-13"><a href="#cb3-13"></a><span class="co">## Create a locus object for the first locus</span></span>
<span id="cb3-14"><a href="#cb3-14"></a>locus =<span class="st"> </span><span class="kw">process.locus</span>(loci[<span class="dv">1</span>,], input)</span></code></pre></div>
<p>If we inspect the locus object, we can see that it contains a lot of info about the locus, such as:</p>
<ul>
<li><p>Locus coordinates (‘chr’,‘start’,‘stop’)</p></li>
<li><p>The number of SNPs (‘K.block’) / PCs (‘K’) within the locus</p></li>
<li><p>The (estimated) PC projected joint SNP effects, <span class="math inline">\(\delta\)</span> (‘delta’; see Methods [<strong>paper ref</strong>])</p></li>
<li><p>The sampling covariance <span class="math inline">\(\sigma^2\)</span> (‘sigma’)</p></li>
<li><p>The genetic covariance matrix <span class="math inline">\(\Omega\)</span> (‘omega’), with corresponding correlation matrix <span class="math inline">\(\Omega^*\)</span> (‘omega.cor’):<span class="math inline">\(\Omega = t(\delta)&#39;\delta / K-\sigma^2\)</span> (Methods, [<strong>paper ref</strong>])</p></li>
</ul>
<p>More information about the return variables can be found in the function documentation</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">str</span>(locus)        <span class="co"># inspect locus object</span></span>
<span id="cb4-2"><a href="#cb4-2"></a><span class="co">#&gt; List of 14</span></span>
<span id="cb4-3"><a href="#cb4-3"></a><span class="co">#&gt;  $ id       : int 100</span></span>
<span id="cb4-4"><a href="#cb4-4"></a><span class="co">#&gt;  $ chr      : int 1</span></span>
<span id="cb4-5"><a href="#cb4-5"></a><span class="co">#&gt;  $ start    : int 113418038</span></span>
<span id="cb4-6"><a href="#cb4-6"></a><span class="co">#&gt;  $ stop     : int 114664387</span></span>
<span id="cb4-7"><a href="#cb4-7"></a><span class="co">#&gt;  $ K        : int 260</span></span>
<span id="cb4-8"><a href="#cb4-8"></a><span class="co">#&gt;  $ K.block  : int 2231</span></span>
<span id="cb4-9"><a href="#cb4-9"></a><span class="co">#&gt;  $ delta    : num [1:260, 1:3] -0.001907 0.002891 -0.005859 0.000558 0.002492 ...</span></span>
<span id="cb4-10"><a href="#cb4-10"></a><span class="co">#&gt;   ..- attr(*, &quot;dimnames&quot;)=List of 2</span></span>
<span id="cb4-11"><a href="#cb4-11"></a><span class="co">#&gt;   .. ..$ : NULL</span></span>
<span id="cb4-12"><a href="#cb4-12"></a><span class="co">#&gt;   .. ..$ : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-13"><a href="#cb4-13"></a><span class="co">#&gt;  $ sigma    : num [1:3, 1:3] 8.92e-06 1.32e-06 1.98e-07 1.32e-06 2.65e-06 ...</span></span>
<span id="cb4-14"><a href="#cb4-14"></a><span class="co">#&gt;   ..- attr(*, &quot;dimnames&quot;)=List of 2</span></span>
<span id="cb4-15"><a href="#cb4-15"></a><span class="co">#&gt;   .. ..$ : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-16"><a href="#cb4-16"></a><span class="co">#&gt;   .. ..$ : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-17"><a href="#cb4-17"></a><span class="co">#&gt;  $ omega    : num [1:3, 1:3] 1.39e-06 -2.07e-07 -4.34e-07 -2.07e-07 4.48e-07 ...</span></span>
<span id="cb4-18"><a href="#cb4-18"></a><span class="co">#&gt;   ..- attr(*, &quot;dimnames&quot;)=List of 2</span></span>
<span id="cb4-19"><a href="#cb4-19"></a><span class="co">#&gt;   .. ..$ : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-20"><a href="#cb4-20"></a><span class="co">#&gt;   .. ..$ : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-21"><a href="#cb4-21"></a><span class="co">#&gt;  $ omega.cor: num [1:3, 1:3] 1 -0.263 -0.426 -0.263 1 ...</span></span>
<span id="cb4-22"><a href="#cb4-22"></a><span class="co">#&gt;   ..- attr(*, &quot;dimnames&quot;)=List of 2</span></span>
<span id="cb4-23"><a href="#cb4-23"></a><span class="co">#&gt;   .. ..$ : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-24"><a href="#cb4-24"></a><span class="co">#&gt;   .. ..$ : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-25"><a href="#cb4-25"></a><span class="co">#&gt;  $ N        : Named num [1:3] 500199 377980 377749</span></span>
<span id="cb4-26"><a href="#cb4-26"></a><span class="co">#&gt;   ..- attr(*, &quot;names&quot;)= chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-27"><a href="#cb4-27"></a><span class="co">#&gt;  $ phenos   : chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-28"><a href="#cb4-28"></a><span class="co">#&gt;  $ binary   : logi [1:3, 1] TRUE FALSE FALSE</span></span>
<span id="cb4-29"><a href="#cb4-29"></a><span class="co">#&gt;   ..- attr(*, &quot;names&quot;)= chr [1:3] &quot;depression&quot; &quot;neuro&quot; &quot;bmi&quot;</span></span>
<span id="cb4-30"><a href="#cb4-30"></a><span class="co">#&gt;  $ snps     : chr [1:2231] &quot;rs2360008&quot; &quot;rs1237670&quot; &quot;rs1235629&quot; &quot;rs1216539&quot; ...</span></span>
<span id="cb4-31"><a href="#cb4-31"></a>?<span class="kw">process.locus</span>()  <span class="co"># check out the function documentation</span></span></code></pre></div>
</div>
<div id="perform-the-univariate-test" class="section level2">
<h2>Perform the univariate test</h2>
<p>If we determine the amount of local genetic signal for all phenotypes of interest, we can filter out unassociated loci and dramatically speed up the local r<sub><em>g</em></sub> analysis. This is done with the univariate test.</p>
<p>After obtaining the locus object, we can pass this object directly to the run.univ() function in order to perform the univariate test for all phenotypes - NOTE: a subset of phenotypes may be analysed by passing an ID vector to the ‘pheno’ argument</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1"></a><span class="kw">run.univ</span>(locus)</span>
<span id="cb5-2"><a href="#cb5-2"></a><span class="co">#&gt;         phen         var          p</span></span>
<span id="cb5-3"><a href="#cb5-3"></a><span class="co">#&gt; 1 depression 1.39200e-06 0.04240950</span></span>
<span id="cb5-4"><a href="#cb5-4"></a><span class="co">#&gt; 2      neuro 4.47714e-07 0.03154340</span></span>
<span id="cb5-5"><a href="#cb5-5"></a><span class="co">#&gt; 3        bmi 7.44367e-07 0.00146622</span></span></code></pre></div>
<p>Here, there was little evidence of any signal for either phenotypes, and there will be little point in testing the local r<sub><em>g</em></sub>.</p>
<p>If we check another locus however:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1"></a>locus =<span class="st"> </span><span class="kw">process.locus</span>(loci[<span class="dv">3</span>,], input)</span>
<span id="cb6-2"><a href="#cb6-2"></a><span class="kw">run.univ</span>(locus)</span>
<span id="cb6-3"><a href="#cb6-3"></a><span class="co">#&gt;         phen         var           p</span></span>
<span id="cb6-4"><a href="#cb6-4"></a><span class="co">#&gt; 1 depression 4.31788e-06 4.13399e-07</span></span>
<span id="cb6-5"><a href="#cb6-5"></a><span class="co">#&gt; 2      neuro 2.13106e-06 1.25993e-14</span></span>
<span id="cb6-6"><a href="#cb6-6"></a><span class="co">#&gt; 3        bmi 3.62544e-06 6.60429e-32</span></span></code></pre></div>
<p>we see that there is plenty of signal for all phenotypes, and we can proceed with the bivariate test</p>
</div>
<div id="bivariate-local-rg-analysis" class="section level2">
<h2>Bivariate local r<sub><em>g</em></sub> analysis</h2>
<p>To proceed with the bivariane rg analysis, we simply pass the locus object the run.bivar() function</p>
<p><strong>TODO: MAKE BIVAR FUNC THAT JUST ANALYSES ALL COMBINATIONS OF PHENOS!!</strong></p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1"></a><span class="kw">run.bivar</span>(locus)</span>
<span id="cb7-2"><a href="#cb7-2"></a><span class="co">#&gt;         pred       rho rho.ci.low rho.ci.high         r2 r2.ci.low r2.ci.high</span></span>
<span id="cb7-3"><a href="#cb7-3"></a><span class="co">#&gt; 1 depression 0.0723039   -0.18291     0.33091 0.00522785         0    0.11306</span></span>
<span id="cb7-4"><a href="#cb7-4"></a><span class="co">#&gt; 2      neuro 0.1069490   -0.10267     0.30983 0.01143810         0    0.09614</span></span>
<span id="cb7-5"><a href="#cb7-5"></a><span class="co">#&gt;          p</span></span>
<span id="cb7-6"><a href="#cb7-6"></a><span class="co">#&gt; 1 0.566464</span></span>
<span id="cb7-7"><a href="#cb7-7"></a><span class="co">#&gt; 2 0.301766</span></span></code></pre></div>
<p>And if we look at the description of this function, we see that this too can take a phenotype vector, should only a subset of phenotypes be analysed. If multiple phenotypes are entered simultaneously, the last phenotype will be treated as the phenotype of interest, and separate bivariate tests will be performed between this phenotype and all others. (Note: this test is symmetric, which phenotype is considered the ‘outcome’ here doesn’t matter).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1"></a>?<span class="kw">run.bivar</span>()</span></code></pre></div>
<p>If we want to analyse the bivariate local r<sub><em>g</em></sub> between a subset phenotypes <strong>[…]</strong></p>
</div>
<div id="obtaining-sample-overlap-from-ldsc" class="section level2">
<h2>Obtaining sample overlap from LDSC</h2>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
